### PR TITLE
refactor(actions): create shared checkout-and-harden composite action

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Summary

Adds a shared `checkout-and-harden` composite action and migrates 65 job preambles across 40 reusable workflows from the duplicated four-step sequence (tooling sparse checkout → `resolve-egress-allowlist` → `harden-runner`) to a bootstrap sparse checkout plus a single `Checkout and harden` step. Net -75 lines with behavior preserved exactly (same sparse-checkout paths per workflow via `sparse-checkout-extra`, same egress inputs, same step ids).

### Spike outcome (pre-hook timing)

The contract docs said step-security's pre-hook "cannot read composite step outputs". The spike found the operative behavior is stronger and simpler:

- The Actions runner **never executes `pre` for workspace-local actions** (`ActionRunner.cs` emits ``"`pre` execution is not supported for local action"``). Live CI logs for this repo confirm it: every job shows that warning and only `[harden-runner] main-step` / `post-step` lines — the pre-hook does not run today either.
- Therefore step-security inputs are evaluated when the step itself runs, and a prior **sibling step inside the same composite** provides `steps.egress.outputs['allowed-endpoints']` with identical timing to the previous workflow-level pattern. No regression is possible via the pre-hook; the docs paragraph was corrected accordingly.
- Nested local refs load just-in-time (`PrepareRepositoryActionAsync` returns early for self-alias refs), so the composite can perform the full tooling checkout itself and then invoke `resolve-egress-allowlist` / `harden-runner` from the directory it just materialized. `prepare-pypi-upload` already relies on the same workspace-relative nesting.

### Design

- Workflow bootstrap checkout materializes only `.github/actions/checkout-and-harden` at the same ref expression as before (`tooling-ref` falling back to `github.workflow_sha`); the composite re-checks-out `.lgtm-ci-tooling` with the base egress paths plus `sparse-checkout-extra`.
- Inputs: `tooling-ref`, `egress-policy` (block), `egress-preset`, `allowed-endpoints`, `allowed-endpoints-mode` (replace), `sparse-checkout-extra`, `persist-credentials` (false). Outputs: `allowed-endpoints`, `scripts-dir`.
- Migration was fully mechanical (scripted transform, re-run verified byte-identical); the one-off script is not committed, matching the repo's removal of the previous `migrate-egress-via-tooling-checkout.py` as dead code.

### Exemptions (unchanged workflows)

- `reusable-release-auto-tag`, `reusable-release-version-pr` — two-phase egress/scripts checkouts around the app token.
- `reusable-build-rust-binaries`, `reusable-publish-rust-release` — `validate-runner-policy` must run between the tooling checkout and resolve/harden (tier gating with `if:` conditions).
- `reusable-validate-lintro-version` — bootstrap/fallback tooling checkout flow.

### Contract check changes (shape genuinely changed)

- `validate-tooling-sparse-checkout.sh`: merges the composite's `sparse-checkout-extra` paths into the checkout block before applying the `scripts/ci/` rule; exempts `checkout-and-harden` alongside the egress composites. New pass/fail BATS cases cover the composite form.
- `validate-harden-runner-action-ref.sh`: new checks for composite users (bootstrap checkout must exist and precede the composite; caller-local egress paths still forbidden). The explicit triplet checks continue to apply to the exempt workflows.
- `egress_tooling_checkout_order_ok` helper and four workflow-specific BATS suites accept the single-step form.

## Type of Change

- [x] New feature (composite action, reusable workflow, or utility script)
- [ ] Bug fix
- [x] Documentation update
- [x] Refactoring (no functional changes)
- [x] CI/CD improvement

## Checklist

- [x] I have tested these changes locally (`bats --recursive tests/bats`: 2354 tests, sole failure is a pre-existing local-env case that also fails on `main`; new composite unit suite passes)
- [x] I have updated documentation if needed (`docs/workflow-contract.md`, `.github/actions/README.md`, `CHANGELOG.md`)
- [x] Shell scripts pass `shellcheck`
- [x] YAML files pass `yamllint` (and `actionlint`)
- [x] Python code passes `ruff` and `black`

## Breaking Changes

None. Reusable workflow inputs and job/step contracts observable by consumers are unchanged; the tooling checkout additionally materializes `.github/actions/checkout-and-harden`.

## Closes

- Closes #379

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds a shared checkout-and-harden workflow preamble. The main changes are:

- New composite action for tooling checkout, egress resolution, and runner hardening.
- Reusable workflows migrated from separate preamble steps to the composite action.
- Contract checks and BATS coverage updated for the new workflow shape.
- Docs updated to describe the composite action and runner timing behavior.
</details>

<h3>Confidence Score: 5/5</h3>

This looks safe to merge.

- No blocking issues found in the changed code.
- The latest validator changes cover the job-boundary cases raised earlier.
- The added tests cover the quoted and indented job forms that previously leaked state.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| scripts/ci/actions/validate-harden-runner-action-ref.sh | Updates the hardening contract validator to reset checkout tracking at job boundaries. |
| .github/actions/checkout-and-harden/action.yml | Adds the shared composite action for tooling checkout, egress resolution, runner hardening, and output wiring. |
| tests/bats/integration/test_harden_runner_action_ref.bats | Adds tests for checkout-and-harden ordering across quoted and indented job keys. |

</details>

<sub>Reviews (4): Last reviewed commit: ["fix(actions): detect job boundary at any..."](https://github.com/lgtm-hq/lgtm-ci/commit/c836ad542926021702e2a2058399c8510d1d5b11) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41930338)</sub>

### CODE REVIEW

File: scripts/ci/actions/validate-harden-runner-action-ref.sh [hunk 1]
```diff
@@ -184,9 +184,40 @@ if grep -qE "^\s+egress-preset:" "$HARDEN_ACTION"; then
...
+	if ! awk '
+		/- name: Checkout lgtm-ci tooling/ { tooling = NR }
+		/uses:[[:space:]]+\.\/\.lgtm-ci-tooling\/\.github\/actions\/checkout-and-harden/ {
+			if (tooling == 0 || tooling >= NR) {
+				bad = 1
+			}
+			tooling = 0
+		}
+		END { exit bad }
+	' "$workflow"; then
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/ci/actions/validate-harden-runner-action-ref.sh:241]: Fixed in 6a6c8a6 — the order check now tracks jobs: scope and resets the tooling line at every job boundary (same pattern as _check_job_egress_order), so a later job using checkout-and-harden without its own bootstrap checkout is flagged. Verified the two-job leak fixture now fails the check.

File: scripts/ci/actions/validate-harden-runner-action-ref.sh [hunk 2]
```diff
@@ -184,9 +188,49 @@ if grep -qE "^\s+egress-preset:" "$HARDEN_ACTION"; then
...
+	if ! awk '
+		BEGIN { in_jobs = 0; tooling = 0 }
+		/^jobs:/ { in_jobs = 1; next }
+		!in_jobs { next }
+		# Job-key boundary: reset the tooling checkout carried from a
+		# previous job. Recognize quoted job keys (e.g. a double- or
+		# single-quoted "release":) as well as bare keys, so a quoted job
+		# cannot inherit a prior job tooling checkout and bypass this
+		# contract (\047 is the octal escape for a single quote).
+		/^  (["\047][^"\047]*["\047]|[a-zA-Z_][a-zA-Z0-9_-]*): *$/ { tooling = 0; next }
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/ci/actions/validate-harden-runner-action-ref.sh]: Fixed in c836ad5. The job-boundary reset no longer assumes two-space indentation: both awk checks now detect the job indent from the first key under `jobs:` and reset carried tooling state at that indent, so a 4-space (or deeper, bare or quoted) job key is recognized as a boundary and cannot inherit a prior job's tooling checkout. Added two discriminating regression tests (4-space leak is flagged; 4-space bootstrapped passes).
